File: IRPlayer-swift/Class/Core/FFPlayer/FFmpeg/Frame/IRFFAudioFrame.m
//
//  IRFFAudioFrame.m
//  IRPlayer
//
//  Created by Phil on 2019/7/5.
//  Copyright © 2019 Phil. All rights reserved.
//

#import "IRFFAudioFrame.h"

@implementation IRFFAudioFrame
{
    size_t buffer_size;
}

- (IRFFFrameType)type
{
    return IRFFFrameTypeAudio;
}

- (int)size
{
    return (int)self->_length;
}

- (void)setSamplesLength:(NSUInteger)samplesLength
{
    if (self->buffer_size < samplesLength) {
        if (self->buffer_size > 0 && self->_samples != NULL) {
            free(self->_samples);
        }
        self->buffer_size = samplesLength;
        self->_samples = malloc(self->buffer_size);
    }
    self->_length = (int)samplesLength;
    self->_output_offset = 0;
}

- (void)dealloc
{
    if (self->buffer_size > 0 && self->_samples != NULL) {
        free(self->_samples);
    }
}

@end

File: IRPlayer-swift/Class/Core/FFPlayer/FFmpeg/Frame/IRFFAudioFrame.swift
//
//  IRFFAudioFrame.swift
//  IRPlayer-swift
//
//  Created by irons on 2024/7/26.
//

import Foundation

@objcMembers public class IRFFAudioFrame: IRFFFrame {
    var samples: UnsafeMutablePointer<Float>?
    var outputOffset: Int = 0
    private var bufferSize: Int = 0

    override var type: IRFFFrameType {
        return .audio
    }

    func setSamplesLength(_ samplesLength: Int) {
        if bufferSize < samplesLength {
            if bufferSize > 0, let samples = samples {
                free(samples)
            }
            bufferSize = samplesLength
            samples = UnsafeMutablePointer<Float>.allocate(capacity: bufferSize)
        }
        size = samplesLength
        outputOffset = 0
    }

    deinit {
        if bufferSize > 0, let samples = samples {
            free(samples)
        }
    }
}

